### PR TITLE
[libc] support PIE relocations

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
For some reasons, we are using `-fpie` (libc/cmake/modules/LLVMLibCObjectRules.cmake:31) without supporting it.
According to @lntue, some of the hermetic tests are broken without proper PIE support. This patch implements basic relocations support for PIE.

### CODE REVIEW

File: libc/startup/linux/do_start.cpp [hunk 1]
```diff
@@ -94,18 +97,26 @@ static ThreadAttributes main_thread_attrib;
...
   for (uintptr_t i = 0; i < program_hdr_count; ++i) {
-    Elf64_Phdr *phdr = program_hdr_table + i;
-    if (phdr->p_type != PT_TLS)
-      continue;
-    // TODO: p_vaddr value has to be adjusted for static-pie executables.
-    app.tls.address = phdr->p_vaddr;
-    app.tls.size = phdr->p_memsz;
-    app.tls.init_size = phdr->p_filesz;
-    app.tls.align = phdr->p_align;
+    auto &phdr = program_hdr_table[i];
```
Comment (jhuber6) [inline libc/startup/linux/do_start.cpp:105]: Nit, explicit type is always better.

File: libc/startup/linux/do_start.cpp [hunk 2]
```diff
@@ -29,6 +29,9 @@ extern uintptr_t __init_array_start[];
 extern uintptr_t __init_array_end[];
 extern uintptr_t __fini_array_start[];
 extern uintptr_t __fini_array_end[];
+// https://refspecs.linuxbase.org/elf/gabi4+/ch5.dynamic.html#dynamic_section
+[[gnu::weak,
```
Comment (jhuber6) [inline libc/startup/linux/do_start.cpp:33]: This is `extern_weak` because we expect it to be undefined and thus null with a PIC executable, right?

**Review by jhuber6 — APPROVED** [review]

File: libc/startup/linux/do_start.cpp [hunk 3]
```diff
@@ -94,18 +99,26 @@ static ThreadAttributes main_thread_attrib;
...
-      continue;
-    // TODO: p_vaddr value has to be adjusted for static-pie executables.
-    app.tls.address = phdr->p_vaddr;
-    app.tls.size = phdr->p_memsz;
-    app.tls.init_size = phdr->p_filesz;
-    app.tls.align = phdr->p_align;
+    Elf64_Phdr &phdr = program_hdr_table[i];
+    if (phdr.p_type == PT_PHDR)
+      base = reinterpret_cast<ptrdiff_t>(program_hdr_table) - phdr.p_vaddr;
+    if (phdr.p_type == PT_DYNAMIC && _DYNAMIC)
```
Comment (nickdesaulniers) [inline libc/startup/linux/do_start.cpp:110]: re: `&& _DYNAMIC` isn't that always true?

Comment (SchrodingerZhu) [reply libc/startup/linux/do_start.cpp:110]: _DYNAMIC is declared as `[[gnu::weak]]`. ELF specifies that undefined weak symbols should give NULL

Comment (SchrodingerZhu) [reply libc/startup/linux/do_start.cpp:110]: "Unresolved weak symbols have a zero value." See also https://maskray.me/blog/2021-04-25-weak-symbol

**Pull request merged**